src/lib/operators/jit_operator/operators/jit_read_tuple.cpp

```diff
@@ -0,0 +1,117 @@
+  for (const auto& input_literal : _input_literals) {
+    auto data_type = input_literal.tuple_value.data_type();
+    resolve_data_type(data_type, [&](auto type) {
+      using DataType = typename decltype(type)::type;
+      input_literal.tuple_value.materialize(ctx).set<DataType>(boost::get<DataType>(input_literal.value));
+    });
+  }
+
+  // Create a JitColumnReader for each input column. We use the first chunk in the table to determine the column type.
+  for (const auto& input_column : _input_columns) {
+    const auto column_id = input_column.column_id;
+    const auto column = in_table.get_chunk(opossum::ChunkID{0})->get_column(column_id);
+    const auto data_type = in_table.column_type(column_id);
+    const auto is_nullable = in_table.column_is_nullable(column_id);
+
+    opossum::resolve_data_and_column_type(data_type, *column, [&](auto type, auto& typed_column) {
+      using ColumnDataType = typename decltype(type)::type;
+      auto iterable = opossum::create_iterable_from_column<ColumnDataType>(typed_column);
+      iterable.with_iterators([&](auto it, auto end) {
+        using IteratorType = decltype(it);
+        if (is_nullable) {
+          _column_readers.push_back(std::make_shared<JitColumnReader<IteratorType, ColumnDataType, true>>(
+              _column_readers.size(), input_column.tuple_value));
+        } else {
+          _column_readers.push_back(std::make_shared<JitColumnReader<IteratorType, ColumnDataType, false>>(
+              _column_readers.size(), input_column.tuple_value));
+        }
+      });
+    });
+  }
+}
+
+void JitReadTuple::before_chunk(const Table& in_table, const Chunk& in_chunk, JitRuntimeContext& ctx) const {
+  ctx.inputs.clear();
+
+  // Create the column iterator for each input column and store them to the runtime context
+  for (const auto& input_column : _input_columns) {
+    const auto column_id = input_column.column_id;
+    const auto column = in_chunk.get_column(column_id);
+    const auto column_data_type = in_table.column_type(column_id);
+    resolve_data_and_column_type(column_data_type, *column, [&](auto type, auto& typed_column) {
+      using ColumnDataType = typename decltype(type)::type;
+      create_iterable_from_column<ColumnDataType>(typed_column).with_iterators([&](auto it, auto end) {
+        using IteratorType = decltype(it);
+        ctx.inputs.push_back(std::make_shared<IteratorType>(it));
+      });
+    });
+  }
+}
+
+void JitReadTuple::execute(JitRuntimeContext& ctx) const {
+  for (; ctx.chunk_offset < ctx.chunk_size; ++ctx.chunk_offset) {
+    // We read from and advance all column iterators, before passing the tuple on to the next operator.
+    for (const auto& column_reader : _column_readers) {
+      column_reader->read_value(ctx);
+      column_reader->increment(ctx);
+    }
+    emit(ctx);
+  }
+}
+
+JitTupleValue JitReadTuple::add_input_column(const Table& table, const ColumnID column_id) {
+  // There is no need to add an input column twice
```

Comment:
How about you wrap that whole block in a `DebugAssert` and throw an exception if someone does?

Reply:
I think the comment is misleading. This function can be called multiple times for the same input column. However, internally, the operator only needs to create the `JitTupleValue` once and can return the same (previously constructed) value on subsequent requests. I updated the comment to clarify.